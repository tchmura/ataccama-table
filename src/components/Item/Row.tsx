import React, { useState } from 'react';
import styled from 'styled-components/macro';

import { Cell } from './Cell';
import { Kids } from '../../types';
import { ItemTable } from './ItemTable';
import { useStores } from '../../hooks/useStores';

type Props = {
  cellsData: string[];
  kids: Kids;
  itemId: string;
  rootColumnCount?: number;
};

const doesKidsHaveRecords = (kids: Kids) => {
  return !!Object.values(kids)
    .map(value => value.records)
    .flat(Infinity).length;
};

export const Row: React.FC<Props> = ({ cellsData, kids, rootColumnCount, itemId }) => {
  const [shouldExpand, setShouldExpand] = useState(false);
  const { itemStore } = useStores();

  const kidsContainerNames = Object.keys(kids);

  return (
    <>
      <StyledRow>
        <StyledButtons>
          <StyledDeleteButton onClick={() => itemStore.deleteItem(itemId)}>delete</StyledDeleteButton>
          {doesKidsHaveRecords(kids) && (
            <button onClick={() => setShouldExpand(prevShouldExpand => !prevShouldExpand)}>
              {shouldExpand ? 'Wrap' : 'Expand'}
            </button>
          )}
        </StyledButtons>
        {cellsData.map((cellText, idx) => (
          <Cell text={cellText} key={idx} />
        ))}
      </StyledRow>
      {doesKidsHaveRecords(kids) && shouldExpand && (
        <tr>
          <td colSpan={rootColumnCount}>
            {kidsContainerNames.map((name, idx) => (
              <ItemTable items={kids[name].records} tableName={name} key={idx} rootColumnCount={rootColumnCount} />
            ))}
          </td>
        </tr>
      )}
    </>
  );
};

const StyledRow = styled.tr`
  border: solid 1px;
`;

const StyledButtons = styled.td`
  display: flex;
  justify-content: space-evenly;

  & button {
    width: 100%;
    border-radius: 5px;
    font-weight: bold;
    margin-right: 10px;
  }

  & button:last-child {
    margin-right: 0;
  }
`;

const StyledDeleteButton = styled.button`
  background-color: red;
  color: white;
`;
