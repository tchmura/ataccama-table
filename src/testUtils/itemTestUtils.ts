import { Kids, Item } from '../types';

export function createKidsData(data: Partial<Kids> = {}): Kids {
  return {
    has_relatives: {
      records: [
        {
          id: 'mockId1',
          data: {
            'Relative ID': '1007',
            'Patient ID': '34',
            'Is alive?': 'true',
            'Frequency of visits': '29'
          },
          kids: {
            has_phone: {
              records: [
                {
                  id: 'mockId1',
                  data: {
                    'Phone ID': '2008',
                    'ID of the relative': '1007',
                    Phone: '+(179)-982-0570'
                  },
                  kids: {}
                }
              ]
            }
          }
        }
      ]
    },
    ...data
  };
}

export function createItemData(data: Partial<Item> = {}): Item {
  return {
    id: 'mockItemId',
    data: {
      'Identification number': '34',
      Name: 'Joqmo',
      Gender: 'female',
      Risk: 'BITES',
      'Hair length': '6.2000000000',
      IQ: '98',
      'Admission date': 'Mon Dec 13 00:00:00 CET 1993',
      'Last breakdown': 'Wed Dec 24 07:14:50 CET 2014',
      'Yearly fee': '67035',
      'Knows the Joker?': 'true'
    },
    kids: { ...createKidsData() },
    ...data
  };
}
