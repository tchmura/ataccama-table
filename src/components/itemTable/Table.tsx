import React from 'react';
import styled from 'styled-components';
import { useObserver } from 'mobx-react';

import { Item } from '../../types';
import { Title } from './Title';
import { Row } from './Row';

type Props = {
  items: Item[];
  tableName?: string;
  rootColumnCount?: number;
};

export const Table: React.FC<Props> = ({ items, tableName }) => {
  return useObserver(() => {
    if (items.length) {
      // assumption - items on the same level have consistent data keys
      const headings = Object.keys(items[0].data);
      const rowsToDisplay = items.map(item => ({ cellsData: Object.values(item.data), itemId: item.id }));
      const kidsList = items.map(item => item.kids);

      return (
        <StyledTable data-testid="item-table" cellPadding={10}>
          <thead>
            <tr>
              <StyledTableName>{tableName}</StyledTableName>
            </tr>
            <StyledHeading>
              <StyledButtonsHeader />
              {headings.map((title, idx) => (
                <Title title={title} key={idx} />
              ))}
            </StyledHeading>
          </thead>
          <tbody>
            {rowsToDisplay.map(({ cellsData, itemId }, idx) => (
              <Row
                cellsData={cellsData}
                kids={kidsList[idx]}
                itemId={itemId}
                rootColumnCount={headings.length + 1 /* +1 to accomodate the column for expand button*/}
                key={itemId}
              />
            ))}
          </tbody>
        </StyledTable>
      );
    }

    return null;
  });
};

const StyledTable = styled.table`
  border-collapse: collapse;
  width: 100%;
  margin-bottom: 20px;
`;

const StyledHeading = styled.tr`
  background-color: lightgray;
  border: solid 1px;
`;

const StyledTableName = styled.td`
  font-size: 25px;
  font-weight: bold;
`;

const StyledButtonsHeader = styled.th`
  /* Explicit width ensures all nested tables have equally wide first column which contains buttons */
  width: 1px;
`;
