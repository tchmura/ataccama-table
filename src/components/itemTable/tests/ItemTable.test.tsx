import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import 'jest-styled-components';

import { doesKidsHaveRecords } from '../Row';
import { createKidsData, createItemData } from '../../../testUtils/itemTestUtils';
import { Table } from '../Table';

describe('doesKidsHaveRecords', () => {
  afterEach(cleanup);

  it('should return true for kids which have records', () => {
    expect(doesKidsHaveRecords(createKidsData())).toEqual(true);
  });

  it('should return false for kids without records', () => {
    expect(doesKidsHaveRecords(createKidsData({ has_relatives: { records: [] } }))).toEqual(false);
  });
});

describe('Table rendering', () => {
  afterEach(cleanup);

  it('should render correctly', () => {
    const { container } = render(<Table items={[createItemData()]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should expand', () => {
    const { getByText, queryByText, queryAllByTestId } = render(<Table items={[createItemData()]} />);

    expect(queryAllByTestId('item-table').length).toEqual(1);
    expect(queryByText('Wrap')).toBeFalsy();

    fireEvent.click(getByText('Expand'));

    expect(queryByText('Wrap')).toBeTruthy();
    expect(queryAllByTestId('item-table').length).toEqual(2);
  });
});
