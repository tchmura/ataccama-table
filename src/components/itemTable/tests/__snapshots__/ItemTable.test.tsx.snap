// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table rendering should render correctly 1`] = `
.c8 {
  border-right: solid 1px;
  border-left: solid 1px;
}

.c4 {
  font-size: 20px;
  font-weight: bold;
  padding: 5px 10px;
}

.c0 {
  border-collapse: collapse;
  width: 100%;
  margin-bottom: 20px;
}

.c2 {
  background-color: lightgray;
  border: solid 1px;
}

.c1 {
  font-size: 25px;
  font-weight: bold;
}

.c3 {
  width: 1px;
}

.c5 {
  border: solid 1px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c6 button {
  width: 100%;
  border-radius: 5px;
  font-weight: bold;
  margin-right: 10px;
}

.c6 button:last-child {
  margin-right: 0;
}

.c7 {
  background-color: red;
  color: white;
}

<table
  cellpadding="10"
  class="c0"
  data-testid="item-table"
>
  <thead>
    <tr>
      <td
        class="c1"
      />
    </tr>
    <tr
      class="c2"
    >
      <th
        class="c3"
      />
      <th
        class="c4"
      >
        Identification number
      </th>
      <th
        class="c4"
      >
        Name
      </th>
      <th
        class="c4"
      >
        Gender
      </th>
      <th
        class="c4"
      >
        Risk
      </th>
      <th
        class="c4"
      >
        Hair length
      </th>
      <th
        class="c4"
      >
        IQ
      </th>
      <th
        class="c4"
      >
        Admission date
      </th>
      <th
        class="c4"
      >
        Last breakdown
      </th>
      <th
        class="c4"
      >
        Yearly fee
      </th>
      <th
        class="c4"
      >
        Knows the Joker?
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="c5"
    >
      <td
        class="c6"
      >
        <button
          class="c7"
        >
          delete
        </button>
        <button>
          Expand
        </button>
      </td>
      <td
        class="c8"
      >
        34
      </td>
      <td
        class="c8"
      >
        Joqmo
      </td>
      <td
        class="c8"
      >
        female
      </td>
      <td
        class="c8"
      >
        BITES
      </td>
      <td
        class="c8"
      >
        6.2000000000
      </td>
      <td
        class="c8"
      >
        98
      </td>
      <td
        class="c8"
      >
        Mon Dec 13 00:00:00 CET 1993
      </td>
      <td
        class="c8"
      >
        Wed Dec 24 07:14:50 CET 2014
      </td>
      <td
        class="c8"
      >
        67035
      </td>
      <td
        class="c8"
      >
        true
      </td>
    </tr>
  </tbody>
</table>
`;
